Refactor API for albums feature

jbaez opened the pull request:
- moved `shared` endpoints to `album`
- make the album API more consistent with a common restful approach and update as needed per spec.
- add validation for all params
- update mobile "shared album service" to use the new API endpoints

@alextran1502:
Let me know what you think of the restful API naming convention. If you like it we could move the other API endpoints to follow the same convention.

You can take a look at implementing the feature in the mobile App if you want. I haven't used Flutter before, but have worked with ReactNative and Swift, so could give it a try if you are working in other things and don't have time.

I haven't created a migration for renaming the tables yet.
Also there are no tests as the test suite needs to be setup properly. I might look at doing this when I have some time. I haven't worked with NestJS and Type ORM before, so would need to find out how to set e2e tests to run the DB in memory.

jbaez committed:
Rename "shared" to "album"

Prepare moving "SharedAlbums" to "Albums"

jbaez committed:
Update server album API endpoints

jbaez committed:
Update mobile app album endpoints

Also add `putRequest` to mobile network.service

jbaez committed:
Add GET album collection filter

- allow to filter by owner = 'mine' | 'their'
- make sharedWithUserIds no longer required when creating an album

jbaez committed:
Rename remaining variables to "album"

jbaez committed:
Add ParseMeUUIDPipe to validate uuid or `me`

jbaez committed:
Add album params validation

jbaez committed:
Update todo in mobile album service.

alextran1502 commented:
Hi @jbaez!

Thank you very much for this PR. I haven't had a chance to thoroughly read through everything but overall seems very good. I agree with the naming convention for the REST Apis and generally, I try to follow the convention. The assets APIs have some exceptions since their use case are more dynamic which is why they don't follow the convention closely. 

I have a few ideas for filtering the `owned` vs `shared` album. Let me draw out some things from the UI point of view and you can also see it better so we can design the APIs for those use cases.

jbaez committed:
Setup e2e testing

jbaez committed:
Add user e2e tests

jbaez committed:
Rename database host env variable to DB_HOST

jbaez committed:
Merge branch 'add/test-setup' into add/albums

jbaez committed:
Add some `Album` e2e tests

Also fix issues found with the tests

alextran1502 commented on server/src/config/database.config.ts:
```diff
@@ -2,7 +2,7 @@ import { TypeOrmModuleOptions } from '@nestjs/typeorm';
 export const databaseConfig: TypeOrmModuleOptions = {
   type: 'postgres',
-  host: 'immich_postgres',
+  host: process.env.HOST || 'immich_postgres',
```
Let's change the `process.env.HOST` to `process.env.DB_HOSTNAME`

jbaez replied:
And also same weird thing happened here with the "ignored" `Rename database host env variable to DB_HOST]` 🤔 . Seems like a "GitHub bug" to me 😅 .
I wonder, maybe if I force push an empty commit in `add/albums` it would "fix it".

jbaez replied:
The forced push in `add/albums` did the trick 🎉

alextran1502 commented on server/test/test-utils.ts:
```diff
@@ -0,0 +1,41 @@
+type CustomAuthCallback = () => AuthUserDto;
+
+export async function initDb() {
+  return getConnection().synchronize(true);
```
Does this action create a database in memory or also on the Postgres container?

I've known some problems that the synchronize feature cannot give the same database structure that we are having using migration. Maybe using migration here instead.

jbaez committed:
Force push (try to recover DB_HOST env)

jbaez committed:
Rename db host env variable to `DB_HOSTNAME`

jbaez committed:
Remove unnecessary `initDb` from test-utils

The current database.config is running the migrations:
`migrationsRun: true`

jbaez committed:
Merge branch 'add/test-setup' into add/albums

jbaez committed:
Remove `initDb` usage from album e2e test

jbaez committed:
Update GET albums filter to `shared`

- add filter by all / shared / not shared
- add response DTOs
- add GET albums e2e tests

jbaez committed:
Merge remote-tracking branch 'origin/main' into add/albums

alextran1502 commented:
@jbaez I like the new change with the boolean value, much clearer on the API level :)

I created a Discord server to discuss the development of the project, there is one in there already. Feel free to jump in https://discord.gg/TvbtmRhN

jbaez committed:
Merge remote-tracking branch 'origin/main' into add/albums

jbaez committed:
Update album e2e tests for user.service changes

jbaez committed:
Update mobile app to use album response DTOs

jbaez committed:
Merge remote-tracking branch 'origin/main' into add/albums

jbaez committed:
Refactor album-service DB into album-registry

- DB logic refactored into album-repository making it easier to test
- add some album-service unit tests
- add `clearMocks` to jest configuration

jbaez committed:
Finish implementing album.service unit tests

jbaez commented:
@alextran1502 This should be ready for review.
- I have refactored all the ORM stuff into a `album-registry` which makes unit-testing much easier and enjoyable 🙂
- There are new unit-tests covering album service (fixed some current issues while writing them). They are based in the current logic, we can update later if needed. (Great as documentation to see what is the expected business logic for albums)
- Album service now return DTOs. The mobile app has been updated to work with the new album response DTOs.
- The album `getList` has been refactored to use Query Builder and new `search` filter
- The endpoints have been updated to standard REST conventions
- There are so many changes showing in git due to the change of file structure and file renames, that it might be easier to checkout the branch and check in the IDE

I think that's all. Let me know if you have any questions 😄

alextran1502 commented on mobile/lib/shared/services/network.service.dart:
```diff
@@ -83,6 +83,25 @@ class NetworkService {
+      debugPrint("DioError: ${e.response}");
+      return null;
+    } catch (e) {
+      debugPrint("ERROR BackupService: $e");
```
This should be changed to "ERROR PutRequest"

jbaez replied:
Sure. I have also changed the rest of `debugPrint` for the other methods in `network.service`, since almost all of them had unrelated text there 👍

alextran1502 commented on server/apps/immich/src/api-v1/album/response-dto/album.ts:
```diff
@@ -0,0 +1,30 @@
+import { mapAsset } from '../../asset/response-dto/asset';
+import { Asset } from '../../asset/response-dto/asset';
+import { User } from '../../user/response-dto/user';
+
```
Can we changed this file name to `album-response.dto.ts` it will make it easier to distinguish when you have dozen tabs open in the IDE or fuzzy search

jbaez replied:
done 👍

alextran1502 commented on server/apps/immich/src/api-v1/album/response-dto/album.ts:
```diff
@@ -0,0 +1,30 @@
+import { Asset } from '../../asset/response-dto/asset';
+import { User } from '../../user/response-dto/user';
+
+export interface Album {
```
Also, can we change the name to `AlbumResponseDto` just to make it explicit?

jbaez replied:
yep 👍

alextran1502 commented on server/apps/immich/src/api-v1/album/album.service.spec.ts:
```diff
@@ -0,0 +1,414 @@
+import { BadRequestException, NotFoundException, ForbiddenException } from '@nestjs/common';
+import { AlbumEntity } from '@app/database/entities/album.entity';
+import { Album } from './response-dto/album';
+
```
I am learning a lot of about testing from reading these - Thank you very much for that!

jbaez replied:
My pleasure. Next is trying TDD (Red - Green - Refactor) 😉

alextran1502 commented on server/apps/immich/src/api-v1/album/album-repository.ts:
```diff
@@ -0,0 +1,228 @@
+import { GetAlbumsDto } from './dto/get-albums.dto';
+import { RemoveAssetsDto } from './dto/remove-assets.dto';
+import { UpdateAlbumDto } from './dto/update-album.dto';
+
```
I believe this pattern is from .NET correct? 

Why do you think it is necessary to go down another layer of obstraction vs the original flow that incorporate the repository/query directly into the `service`? I am new to design the API system so I am here to learn :)

jbaez replied:
It's all about decoupling "layers", which among other things makes testing easier. These decoupling concepts are part of architectures like Onion, Clean, Hexagonal.
Immich is mostly Onion architecture

alextran1502 commented on server/apps/immich/src/api-v1/asset/response-dto/asset.ts:
```diff
@@ -0,0 +1,39 @@
+import { AssetEntity, AssetType } from '@app/database/entities/asset.entity';
+import { Exif, mapExif } from './exif';
+import { SmartInfo, mapSmartInfo } from './smart-info';
+
```
Let's change the name of the files in this folder as well as the class name to explicitly telling the user what they are?

jbaez replied:
👍

alextran1502 commented on server/libs/database/src/entities/album.entity.ts:
```diff
@@ -0,0 +1,27 @@
+import { AssetAlbumEntity } from './asset-album.entity';
+import { UserAlbumEntity } from './user-album.entity';
+
+@Entity('shared_albums')
```
We will need to create migration file for this as well

jbaez replied:
Yep, I left that to the end, and then I forgot to do it 😅 . I have created a migration file including the rename of `shared_albums` to `albums` and `asset_shared_album` to `asset_album`

alextran1502 commented on server/libs/database/src/entities/shared-album.entity.ts:
```diff
@@ -1,27 +0,0 @@
-import { AssetSharedAlbumEntity } from './asset-shared-album.entity';
-import { UserSharedAlbumEntity } from './user-shared-album.entity';
-
-@Entity('shared_albums')
```
Maybe easier (or harder?) to change the table name from DB migration?

alextran1502 reviewed (commented):
Thank you for this PR, it is a lot of work and I appreciate your help! I left some comments, looking forward to your responses.

jbaez committed:
Rename response DTO

Make them consistent with rest of the project naming

jbaez committed:
Update debug log messages in mobile network service

jbaez committed:
Rename table `shared_albums` to `albums`

jbaez committed:
Rename table `asset_shared_album`

jbaez committed:
Rename Albums `sharedAssets` to `assets`

jbaez commented on mobile/lib/modules/sharing/views/album_viewer_page.dart:
```diff
@@ -39,7 +39,7 @@ class AlbumViewerPage extends HookConsumerWidget {
       if (albumInfo.sharedAssets != null && albumInfo.sharedAssets!.isNotEmpty) {
         ref
             .watch(assetSelectionProvider.notifier)
-            .addNewAssets(albumInfo.sharedAssets!.map((e) => e.assetInfo).toList());
```
Yep, before the album related resources were the DB representation of it, so `sharedAssets` was the pivot table to assets table (`AssetAlbumEntity`), that's why there was a map there to return `assetInfo` which was `AssetEntity`.

Now the `sharedAssets` property (note that I have now renamed this to `assets` in the last PR update) is the `AssetResponseDto` (no pivot table returned in the DTO) so there is no need to use map 🙂

jbaez commented:
> Thank you for this PR, it is a lot of work and I appreciate your help! I left some comments, looking forward to your responses.

Happy to help.
I have updated based on the comments.
Let me know if you have questions 🙂

jbaez committed:
Update tests to match updated "delete" response

alextran1502 commented:
I've just tested all the functionality and all works except the ability to show added assets when we choose to add more assets to the existing album. Can you help me check it?

alextran1502 committed:
Fixed asset cannot be compared in Set by adding Equatable package

alextran1502 committed:
Remove hero effect to fixed janky animation
